### PR TITLE
feat: add auth middleware hook to nextjs client

### PR DESCRIPTION
**Pull request by Pagebakers**
This PR adds improvements to the Next.js client.

- Exported useWunderGraphContext from package.
- New useAuthMiddleware helper hook, that allows you to inject tokens into the client.
- withWunderGraph now supports SWR middleware

### CODE REVIEW

**Review by jensneuse — CHANGES_REQUESTED** [review]

File: packages/nextjs/src/auth-middleware.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+	 * Returning null will unset the token header.
+	 */
+	getToken: () => Promise<string | null | undefined>
+) => MiddlewareReturn;
+
+export const useAuthMiddleware: AuthMiddleware = (useSWRNext, getToken) => {
+	return (key, fetcher, config) => {
+		const context = useWunderGraphContext();
+
+		const fetcherWithAuth: BareFetcher<Promise<unknown>> = async (...args) => {
```
[the remark this answers is not in the document]
Comment (Pagebakers) [reply packages/nextjs/src/auth-middleware.ts:19]: No this is not related to auth just a 'bare' fetcher function.

File: packages/nextjs/src/auth-middleware.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+			try {
+				const token = await getToken();
+				if (token) {
+					context?.client.setAuthorizationToken(token);
+				} else if (token === null) {
+					context?.client.unsetAuthorization();
+				}
+			} catch (e: any) {
+				console.error(e);
```
[the remark this answers is not in the document]
Comment (Pagebakers) [reply packages/nextjs/src/auth-middleware.ts:28]: I didn't want to swallow the error, but also not make the entire request fail if the token function throws. What do you think?

[the remark this answers is not in the document]
Comment (fiam) [reply packages/nextjs/src/auth-middleware.ts:28]: What about letting it throw and allowing users to take the measures the deem appropriate? e.g. plugging in a library that logs all exceptions

[the remark this answers is not in the document]
Comment (Pagebakers) [reply packages/nextjs/src/auth-middleware.ts:28]: Fair enough

**Review by fiam — APPROVED** [review]
LGTM

**Review by jensneuse — APPROVED** [review]

**Pull request merged**